hr_timesheet_sheet/models/hr_timesheet_sheet.py

```diff
@@ -498,21 +695,55 @@ def action_timesheet_confirm(self):
     def action_timesheet_done(self):
         if self.filtered(lambda sheet: sheet.state != 'confirm'):
             raise UserError(_('Cannot approve a non-submitted sheet.'))
-        if not self.env.user.has_group('hr_timesheet.group_hr_timesheet_user'):
-            raise UserError(
-                _('Only an HR Officer or Manager can approve sheets.'))
+        self._check_can_review()
         self.write({
             'state': 'done',
+            'reviewer_id': self._get_current_reviewer().id,
         })
 
     @api.multi
     def action_timesheet_refuse(self):
         if self.filtered(lambda sheet: sheet.state != 'confirm'):
             raise UserError(_('Cannot reject a non-submitted sheet.'))
+        self._check_can_review()
         self.write({
             'state': 'draft',
+            'reviewer_id': False,
         })
 
+    @api.model
+    def _get_current_reviewer(self):
+        reviewer = self.env['hr.employee'].search(
+            [('user_id', '=', self.env.uid)],
+            limit=1
+        )
+        if not reviewer:
+            raise UserError(_(
+                'In order to review a timesheet sheet, your user needs to be'
+                ' linked to an employee.'
+            ))
+        return reviewer
+
+    @api.multi
+    def _check_can_review(self):
+        for sheet in self.filtered(lambda x: not x.can_review):
```

Comment:
@mreficent @astirpe to give you better insights: this method has another 12 lines of code specific to validation models.